internal/npm_install/generate_build_file.ts

```diff
@@ -0,0 +1,1140 @@
+  log_verbose(`could not find entry point for npm package ${pkg._name}`);
+
+  // at this point there's nothing left for us to try, so return nothing
+  return undefined;
+}
+
+type Bag<T> =
+    {
+      [k: string]: T
+    }
+
+/**
+ * Flattens all transitive dependencies of a package
+ * into a _dependencies array.
+ */
+function flattenPkgDependencies(pkg: Dep, dep: Dep, pkgsMap: Map<string, Dep>) {
+  if (pkg._dependencies.indexOf(dep) !== -1) {
+    // circular dependency
+    return;
+  }
+  pkg._dependencies.push(dep);
+  const findDeps = function(targetDeps: Bag<string>, required: boolean, depType: string) {
+    Object.keys(targetDeps || {})
+        .map(targetDep => {
+          // look for matching nested package
+          const dirSegments = dep._dir.split('/');
+          while (dirSegments.length) {
+            const maybe = path.posix.join(...dirSegments, 'node_modules', targetDep);
+            if (pkgsMap.has(maybe)) {
+              return pkgsMap.get(maybe);
+            }
+            dirSegments.pop();
+          }
+          // look for matching root package
+          if (pkgsMap.has(targetDep)) {
+            return pkgsMap.get(targetDep);
+          }
+          // dependency not found
+          if (required) {
+            console.error(`could not find ${depType} '${targetDep}' of '${dep._dir}'`);
+            process.exit(1);
+          }
+          return null;
+        })
+        .filter(dep => !!dep)
+        .forEach(dep => flattenPkgDependencies(pkg, dep!, pkgsMap));
+  };
+  // npm will in some cases add optionalDependencies to the list
+  // of dependencies to the package.json it writes to node_modules.
+  // We delete these here if they exist as they may result
+  // in expected dependencies that are not found.
+  if (dep.dependencies && dep.optionalDependencies) {
+    Object.keys(dep.optionalDependencies).forEach(optionalDep => {
+      delete dep.dependencies[optionalDep];
+    });
+  }
+
+  findDeps(dep.dependencies, true, 'dependency');
+  findDeps(dep.peerDependencies, true, 'peer dependency');
+  // `optionalDependencies` that are missing should be silently
+  // ignored since the npm/yarn will not fail if these dependencies
+  // fail to install. Packages should handle the cases where these
+  // dependencies are missing gracefully at runtime.
+  // An example of this is the `chokidar` package which specifies
+  // `fsevents` as an optionalDependency. On OSX, `fsevents`
+  // is installed successfully, but on Windows & Linux, `fsevents`
+  // fails to install and the package will not be present when
+  // checking the dependencies of `chokidar`.
+  findDeps(dep.optionalDependencies, false, 'optional dependency');
+}
+
+/**
+ * Reformat/pretty-print a json object as a skylark comment (each line
+ * starts with '# ').
+ */
+function printJson(pkg: Dep) {
+  // Clone and modify _dependencies to avoid circular issues when JSONifying
+  // & delete _files array
+  const cloned: any = {...pkg};
```

Comment:
cloned looks like a `Dep`

Reply:
yeah. i wish. it turns out its cloned._dependencies is a string array not a dep array.
because we only use it in debugging within this method i felt it was more maintainable to use any rather than extend Dep

Reply:
ah... Right, well that's unfortunate.
Might be worth refactoring when looking at this afterwards then